### PR TITLE
Move part of the regression integration tests from scripts repo

### PR DESCRIPTION
**Pull request by arwer13**
Migrated same-named tests from https://github.com/lidofinance/scripts/:
staking-limits.integration.ts
hash-consensus.integration.ts
staking-module.integration.ts
dsm-keys-unvetting.integration.t
withdrawal-edge-cases.integration.ts
dsm-pause-deposits.integration.ts
node-operators-flow.integration.ts
withdrawal-happy-path.integration.ts
el-rewards.integration.ts

NB: the tests have been adopted for the https://github.com/lidofinance/core/pull/978 as well

### CODE REVIEW

File: test/integration/core/dsm-keys-unvetting.integration.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+import { DepositSecurityModule } from "typechain-types";
+
+import { certainAddress, DSMUnvetMessage, ether, impersonate } from "lib";
+import { getProtocolContext, ProtocolContext } from "lib/protocol";
+import { setSingleGuardian } from "lib/protocol/helpers/dsm";
+import { norAddNodeOperator, norAddOperatorKeys, norSetOperatorStakingLimit } from "lib/protocol/helpers/nor";
+
+import { Snapshot } from "test/suite";
+
+const GUARDIAN_PRIVATE_KEY = "0x516b8a7d9290502f5661da81f0cf43893e3d19cb9aea3c426cfb36e8186e9c09";
```
Comment (tamtamchik) [inline test/integration/core/dsm-keys-unvetting.integration.ts:18]: If this is a verifiable constant, it's better to use constants and include a comment above on how to verify it. Otherwise, just add a comment indicating that we are using an arbitrary address.

File: test/integration/core/el-rewards.integration.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import { ethers } from "hardhat";
+
+import { HardhatEthersSigner } from "@nomicfoundation/hardhat-ethers/signers";
+
+import { ether } from "lib";
+import { getProtocolContext, ProtocolContext } from "lib/protocol";
+
+import { Snapshot } from "test/suite";
+
+describe("EL rewards distribution", () => {
```
Comment (tamtamchik) [inline test/integration/core/el-rewards.integration.ts]:
```suggestion
describe("Scenario: EL rewards distribution", () => {
```

Comment (tamtamchik) [reply test/integration/core/el-rewards.integration.ts]: Looks like tests depend on each-other

File: test/integration/core/el-rewards.integration.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  const REWARD_AMOUNT = ether("1");
+
+  before(async () => {
+    ctx = await getProtocolContext();
+    [stranger] = await ethers.getSigners();
+    snapshot = await Snapshot.take();
+  });
+
+  after(async () => await Snapshot.restore(snapshot));
```
Comment (tamtamchik) [inline test/integration/core/el-rewards.integration.ts:25]:
```suggestion
  beforeEach(bailOnFailure);

  after(async () => await Snapshot.restore(snapshot));
```

Not to continue scenario test if one of the tests failed

File: test/integration/core/dsm-keys-unvetting.integration.ts (hunk 1 above)
Comment (tamtamchik) [reply test/integration/core/dsm-keys-unvetting.integration.ts:18]: I see this key is reused multiple times

File: test/integration/core/dsm-pause-deposits.integration.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    const guardianSigner = await impersonate(guardian, ether("1"));
+
+    const blockNumber = await time.latestBlock();
+    await pauseDeposits(guardianSigner, BigInt(blockNumber), { r: ZeroHash, vs: ZeroHash }, guardian);
+    await ownerUnpauseDeposits();
+  });
+
+  it("Should allow stranger to pause deposits with guardian signature", async () => {
+    // Create new guardian with known private key
+    const guardianPrivateKey = "0x516b8a7d9290502f5661da81f0cf43893e3d19cb9aea3c426cfb36e8186e9c09";
```
Comment (tamtamchik) [inline test/integration/core/dsm-pause-deposits.integration.ts:88]: Could we reuse the key somehow instead of using a hardcode in test? (see above)

File: test/integration/core/dsm-keys-unvetting.integration.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+      Number(nonce),
+      nodeOperatorIds,
+      vettedSigningKeysCounts,
+    );
+    const sig = await unvetMessage.sign(GUARDIAN_PRIVATE_KEY);
+
+    // Stranger should be able to unvet with valid guardian signature
+    await dsm
+      .connect(stranger)
+      .unvetSigningKeys(blockNumber, blockHash, stakingModuleId, nonce, nodeOperatorIds, vettedSigningKeysCounts, sig);
```
Comment (tamtamchik) [inline test/integration/core/dsm-keys-unvetting.integration.ts:146]: Should we verify the state **after** the unveting?

Comment (tamtamchik) [reply test/integration/core/dsm-keys-unvetting.integration.ts:146]: Also, validate events?

File: test/integration/core/dsm-keys-unvetting.integration.ts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    // Pack operator IDs into bytes (8 bytes per ID)
+    const nodeOperatorIds = ethers.solidityPacked(["uint64"], [operatorId]);
+
+    // Pack vetted signing keys counts into bytes (16 bytes per count)
+    const vettedSigningKeysCounts = ethers.solidityPacked(["uint128"], [1]);
+
+    // Guardian should be able to unvet directly without signature
+    await dsm
+      .connect(guardianSigner)
+      .unvetSigningKeys(blockNumber, blockHash, stakingModuleId, nonce, nodeOperatorIds, vettedSigningKeysCounts, {
```
Comment (tamtamchik) [inline test/integration/core/dsm-keys-unvetting.integration.ts:192]: Validate events?

File: test/integration/core/dsm-keys-unvetting.integration.ts [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        r: ZeroHash,
+        vs: ZeroHash,
+      });
+
+    // Verify node operator state after unvetting
+    const nodeOperatorsAfter = await nor.getNodeOperator(operatorId, true);
+    expect(nodeOperatorsAfter.totalDepositedValidators).to.equal(totalDepositedValidatorsBefore);
+    expect(nodeOperatorsAfter.totalVettedValidators).to.equal(totalDepositedValidatorsBefore);
```
Comment (tamtamchik) [inline test/integration/core/dsm-keys-unvetting.integration.ts]: Why do we check for equality? Shouldn't the key numbers be different (at least vetted) if we unvet them?

File: test/integration/core/dsm-keys-unvetting.integration.ts [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    // Set single guardian
+    await setSingleGuardian(ctx, stranger.address);
+
+    // Guardian should be able to unvet directly without signature
+    await dsm
+      .connect(stranger)
+      .unvetSigningKeys(blockNumber, blockHash, stakingModuleId, nonce, nodeOperatorIds, vettedSigningKeysCounts, {
+        r: ZeroHash,
+        vs: ZeroHash,
+      });
```
Comment (tamtamchik) [inline test/integration/core/dsm-keys-unvetting.integration.ts:248]: Events?

File: test/integration/core/dsm-keys-unvetting.integration.ts [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+    // Verify node operator state after unvetting
+    const nodeOperatorAfterUnvetting = await nor.getNodeOperator(operatorId, true);
+    expect(nodeOperatorAfterUnvetting.totalVettedValidators).to.equal(vettedSigningKeysCountsAfterUnvet);
```
Comment (tamtamchik) [inline test/integration/core/dsm-keys-unvetting.integration.ts:259]: This is because it used to be 8 and unvetted 4? So 8 - 4 = 4? I would suggest a less complicated assertion, even hardcode with a comment would be easier, imho

File: test/integration/core/dsm-pause-deposits.integration.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    expect(depositsPausedEvents[0].args.guardian).to.equal(guardian);
+    expect(await dsm.isDepositsPaused()).to.be.true;
+
+    return pauseDepositTx;
+  }
+
+  async function ownerUnpauseDeposits() {
+    expect(await dsm.isDepositsPaused()).to.be.true;
+    const owner = await dsm.getOwner();
+    const ownerSigner = await impersonate(owner);
```
Comment (tamtamchik) [inline test/integration/core/dsm-pause-deposits.integration.ts]: No funding? Can't lead to revert because of no funds?

File: test/integration/core/hash-consensus.integration.ts [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+  ProtocolContext,
+  report,
+  waitNextAvailableReportTime,
+} from "lib/protocol";
+
+import { Snapshot, ZERO_HASH } from "test/suite";
+
+import { HashConsensus } from "../../../typechain-types";
+
+describe("Hash consensus negative scenarios", () => {
```
Comment (tamtamchik) [inline test/integration/core/hash-consensus.integration.ts:23]:
```suggestion
describe("Scenario: Hash consensus negative scenarios", () => {
```

Comment (tamtamchik) [reply test/integration/core/hash-consensus.integration.ts:23]: Right now, tests are not isolated, but lit looks like they may be isolated, and then we may use beforeEach / afterEach snapshots

File: test/integration/core/hash-consensus.integration.ts [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    const requiredQuorum = Math.floor((members.length + 1) / 2) + 1;
+    await expect(hashConsensus.connect(stranger).addMember(stranger.address, requiredQuorum - 1))
+      .to.be.revertedWithCustomError(hashConsensus, "QuorumTooSmall")
+      .withArgs(requiredQuorum, requiredQuorum - 1);
+  });
+
+  it("Should validate member removal", async () => {
+    // Grant MANAGE_MEMBERS_AND_QUORUM_ROLE to stranger
+    const MANAGE_MEMBERS_AND_QUORUM_ROLE = await hashConsensus.MANAGE_MEMBERS_AND_QUORUM_ROLE();
+    await hashConsensus.connect(agent).grantRole(MANAGE_MEMBERS_AND_QUORUM_ROLE, stranger.address);
```
Comment (tamtamchik) [inline test/integration/core/hash-consensus.integration.ts:119]: Already granted? Because no snapshot reset between the tests

File: test/integration/core/hash-consensus.integration.ts [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+    const reportHash = calcReportDataHash(items);
+
+    // Test cannot submit with invalid slot
+    await expect(
+      hashConsensus.connect(memberSigner).submitReport(0n, reportHash, consensusVersion),
+    ).to.be.revertedWithCustomError(hashConsensus, "InvalidSlot");
+
+    // Test numeric overflow
+    await expect(
+      hashConsensus.connect(memberSigner).submitReport(2n ** 64n, reportHash, consensusVersion),
```
Comment (tamtamchik) [inline test/integration/core/hash-consensus.integration.ts]: Constant MAX_UINT64?

File: test/integration/core/node-operators-flow.integration.ts [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+      );
+  });
+
+  beforeEach(async () => (originalState = await Snapshot.take()));
+
+  afterEach(async () => await Snapshot.restore(originalState));
+
+  after(async () => await Snapshot.restore(snapshot));
+
+  it("Should allow adding a node operator", async () => {
```
Comment (tamtamchik) [inline test/integration/core/node-operators-flow.integration.ts]: Maybe split into several logical parts and convert to scenario test, to make it simple to understand what has blown up (in case test failed)

File: test/integration/core/node-operators-flow.integration.ts [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+  expect(summaryAfter.totalExitedValidators).to.equal(summaryBefore.totalExitedValidators);
+  expect(summaryAfter.totalDepositedValidators).to.equal(summaryBefore.totalDepositedValidators);
+  if (expectedChanges.depositableValidatorsCount !== undefined) {
+    expect(summaryAfter.depositableValidatorsCount).to.equal(expectedChanges.depositableValidatorsCount);
+  } else {
+    expect(summaryAfter.depositableValidatorsCount).to.equal(summaryBefore.depositableValidatorsCount);
+  }
+}
+
+describe("Integration: Node operators flow", () => {
```
Comment (tamtamchik) [inline test/integration/core/node-operators-flow.integration.ts]: Looks like only happy path, and no edge cases checks

File: test/integration/core/staking-limits.integration.ts [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+  before(async () => {
+    ctx = await getProtocolContext();
+    lido = ctx.contracts.lido;
+    [stranger] = await ethers.getSigners();
+    voting = await ctx.getSigner("voting");
+
+    snapshot = await Snapshot.take();
+  });
+
+  after(async () => await Snapshot.restore(snapshot));
```
Comment (tamtamchik) [inline test/integration/core/staking-limits.integration.ts:39]: Does not look like scenario tests, can use beforeEach/afterEach snapshots

File: test/integration/core/staking-limits.integration.ts [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+  context("Staking limit events", () => {
+    it("Should emit correct event when setting staking limit", async () => {
+      const maxLimit = ether("1000");
+      const limitPerBlock = ether("100");
+
+      const tx = await lido.connect(voting).setStakingLimit(maxLimit, limitPerBlock);
+      const receipt = await tx.wait();
+
+      expect(receipt?.logs.length).to.equal(1);
+      await expect(tx).to.emit(lido, "StakingLimitSet").withArgs(maxLimit, limitPerBlock);
```
Comment (tamtamchik) [inline test/integration/core/staking-limits.integration.ts]: Maybe `await expect(lido.connect(voting).setStakingLimit(maxLimit, limitPerBlock)).to.emit...` we don't use tx here except for events checking, but can be simplified

Comment (tamtamchik) [reply test/integration/core/staking-limits.integration.ts]: And below also

**Review by tamtamchik — APPROVED** [review]
Epic work! 👏 

Added some suggestions, though. feel free to do nothing about them and just merge the tests into `feat/vaults`

File: test/integration/core/dsm-keys-unvetting.integration.ts (hunk 1 above)
Comment (arwer13) [reply test/integration/core/dsm-keys-unvetting.integration.ts:18]: It's arbitrary, added comments

File: test/integration/core/dsm-keys-unvetting.integration.ts (hunk 5 above)
Comment (arwer13) [reply test/integration/core/dsm-keys-unvetting.integration.ts:146]: Indeed, added (it was missing originally)>

File: test/integration/core/dsm-keys-unvetting.integration.ts (hunk 6 above)
Comment (arwer13) [reply test/integration/core/dsm-keys-unvetting.integration.ts:192]: added

File: test/integration/core/dsm-keys-unvetting.integration.ts (hunk 7 above)
Comment (arwer13) [reply test/integration/core/dsm-keys-unvetting.integration.ts]: the comparison with total deposited is incorrect indeed (it was originally) - it accidentally passes for the current constant, removed

File: test/integration/core/dsm-keys-unvetting.integration.ts (hunk 8 above)
Comment (arwer13) [reply test/integration/core/dsm-keys-unvetting.integration.ts:248]: added

File: test/integration/core/dsm-keys-unvetting.integration.ts (hunk 9 above)
Comment (arwer13) [reply test/integration/core/dsm-keys-unvetting.integration.ts:259]: It actually sets it not decreases by the provided count:

https://github.com/lidofinance/core/blob/76626192461a8c41c74203303c0aa24ad8f521b6/contracts/0.4.24/nos/NodeOperatorsRegistry.sol#L478-L480

The test passes not only for 4, but anyway changed it to 3.

File: test/integration/core/dsm-pause-deposits.integration.ts (hunk 4 above)
Comment (arwer13) [reply test/integration/core/dsm-pause-deposits.integration.ts:88]: It's an arbitrary value and used only in two test file

File: test/integration/core/hash-consensus.integration.ts (hunk 11 above)
Comment (arwer13) [reply test/integration/core/hash-consensus.integration.ts:23]: They may be isolated indeed, fixed!

File: test/integration/core/hash-consensus.integration.ts (hunk 12 above)
Comment (arwer13) [reply test/integration/core/hash-consensus.integration.ts:119]: no longer relevant, because tests have been isolated

File: test/integration/core/hash-consensus.integration.ts (hunk 13 above)
Comment (arwer13) [reply test/integration/core/hash-consensus.integration.ts]: fixed

File: test/integration/core/node-operators-flow.integration.ts (hunk 15 above)
Comment (arwer13) [reply test/integration/core/node-operators-flow.integration.ts]: indeed

File: test/integration/core/node-operators-flow.integration.ts (hunk 14 above)
Comment (arwer13) [reply test/integration/core/node-operators-flow.integration.ts]: done

File: test/integration/core/staking-limits.integration.ts (hunk 16 above)
Comment (arwer13) [reply test/integration/core/staking-limits.integration.ts:39]: yes!
